examples/qaoa.lisp

```diff
@@ -0,0 +1,129 @@
+;;;; qaoa.lisp
+;;;;
+;;;; Author: Robert Smith
+
+(in-package #:qvm-examples)
+
+(defun qaoa (p cost-ham)
+  "Produce a QAOA program for the list of Pauli (CL-QUIL.CLIFFORD:PAULI) terms COST-HAM representing a cost Hamiltonian to optimize. Return three values:
+
+    1. The Quil program (as a PARSED-PROGRAM)
+
+    2. A list of gamma parameters (as MEMORY-REFs) which parameterize the cost unitary.
+
+    3. A list of beta parameters (as MEMORY-REFs) which parameterize the driver unitary.
+
+COST-HAM should consist of commuting Pauli terms only. (This is *not* checked.)"
+  (check-type p (integer 1))
+  ;; Every Pauli in COST-HAM should have the same number of qubits.
+  (let* ((n (cl-quil.clifford:num-qubits
+             (alexandria:extremum cost-ham #'> :key #'cl-quil.clifford:num-qubits)))
+         (quil (quil:parse-quil (format nil "DECLARE beta REAL[~D]; DECLARE gamma REAL[~D]" p p)))
+         ;; Parameters
+         (betas (loop :for i :below p :collect (quil:mref "beta" i)))
+         (gammas (loop :for i :below p :collect (quil:mref "gamma" i)))
+         ;; Hamiltonian
+         (driver-ham (loop :for q :below n
+                           :collect (cl-quil.clifford:embed cl-quil.clifford:+X+ n (list q))))
+         (isns (quil:with-inst ()
```

Comment:
`with-inst` is not external

Reply:
it should be in the latest cl-quil

Reply:
should we bump the cl-quil dependency to 1.13.1 then?

Reply:
qvm-examples are low stakes, but prob reasonable